### PR TITLE
Git SSH Custom Port 

### PR DESCRIPTION
**Pull request by nicholasSUSE**
## Issue: 

- https://github.com/rancher/rancher/issues/42806

 
## Problem
The client informed us that it was not possible for Rancher Mapps to clone a repository from a Git SSH Server configured with a port different than the default one (22).

## Solution
Changed the regexp to accept the following URL pattern:
`ssh://<user>@<mydomain.exampl>:<port>/<path-to-repository-on-server>/<repository-name>`

## Testing

To test ideally:
1. Create a custom git SSH server
2. Set up a security group with keys
3. Change default SSH port using Custom TCP at security group
4. Change the default SSH port on the Linux instance
5. Try to clone

### Automated Testing
To be implemented

## QA Testing Considerations

Any SSH URL in the following pattern should work:
`ssh://<user>@<mydomain.exampl>:<port>/<path-to-repository-on-server>/<repository-name>`
 
### Regressions Considerations
Please make sure that the new implementation does not break the most simple SSH URL pattern at least for Git Hub.
example: `git@github.com:nicholasSUSE/helm-charts-test.git`

### CODE REVIEW

**Review by rohitsakala — CHANGES_REQUESTED** [review]
Upon testing, 

1. Adding a repo with SSH along with secrets doesn't work. 
2. Adding a repo without ssh keys doesn't show up error quickly but logs show error quickly

File: pkg/catalogv2/git/utils.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	return valid
+}
+
+// validateGitURL will validate if the provided URL is in one of the expected patterns
+// for the supported protocols http(s) or ssh.
+//   - if Valid: returns nil
+//   - if Invalid: returns an error
+func validateGitURL(gitURL string) error {
```
Comment (rohitsakala) [inline pkg/catalogv2/git/utils.go:67]: Should we change this function name from validateGitURL to validateURL ?

File: pkg/catalogv2/git/utils.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+	valid := isGitSSH(gitURL)
+	if valid {
+		return nil
+	}
+	// not ssh; validate http(s)
+	u, err := url.Parse(gitURL)
+	if err != nil || (u.Scheme != "http" && u.Scheme != "https") {
+		return fmt.Errorf("invalid git URL: %s", gitURL)
```
Comment (rohitsakala) [inline pkg/catalogv2/git/utils.go:75]: The error message should say invalid URL right?

File: pkg/catalogv2/git/download.go [hunk 3]
```diff
@@ -1,100 +1,111 @@
...
-	}
-	if !isGitSSH {
-		u, err := url.Parse(gitURL)
-		if err != nil {
-			return nil, fmt.Errorf("failed to parse URL %s: %w", gitURL, err)
-		}
-		if u.Scheme != "http" && u.Scheme != "https" {
-			return nil, fmt.Errorf("invalid git URL scheme %s, only http(s) and git supported", u.Scheme)
-		}
+		return nil, fmt.Errorf("%w: only http(s) or ssh supported", err)
```
Comment (rohitsakala) [inline pkg/catalogv2/git/download.go:106]: Can we change this to `%w: only http(s) or ssh:// supported`

File: pkg/catalogv2/git/utils.go (hunk 2 above)
Comment (nicholasSUSE) [reply pkg/catalogv2/git/utils.go:75]: OK

File: pkg/catalogv2/git/download.go (hunk 3 above)
Comment (nicholasSUSE) [reply pkg/catalogv2/git/download.go]: OK

File: pkg/catalogv2/git/download.go [hunk 4]
```diff
@@ -82,9 +82,13 @@ func Update(secret *corev1.Secret, namespace, name, gitURL, branch string, insec
 		return commit, fmt.Errorf("update failure: %w", err)
 	}
 
-	if changed, err := git.remoteSHAChanged(branch, commit); err != nil || !changed {
+	changed, err := git.remoteSHAChanged(branch, commit)
```
Comment (rohitsakala) [inline pkg/catalogv2/git/download.go:84]: @nicholasSUSE Can we stick to the previous implementation and make another issue and then PR for this mistake ?

Thanks 😄

**Review by rohitsakala — CHANGES_REQUESTED** [review]

File: pkg/catalogv2/git/utils.go (hunk 2 above)
Comment (nicholasSUSE) [reply pkg/catalogv2/git/utils.go:75]: @rohitsakala please resolve this

File: pkg/catalogv2/git/download.go (hunk 4 above)
Comment (nicholasSUSE) [reply pkg/catalogv2/git/download.go:84]: I am not really sure what you are talking about.

Do you mean the change from Ensure(commit) -> Ensure(branch)?
This is still the Ensure(commit), I am not changing this behavior at this PR.

**Review by rohitsakala — APPROVED** [review]

**Pull request merged**